langchain/src/vectorstores/googlevertexai.ts

```diff
@@ -0,0 +1,682 @@
+      upsertClientParams,
+      this.caller
+    );
+  }
+
+  _vectorstoreType(): string {
+    return "googlevertexai";
+  }
+
+  async addDocuments(documents: Document[]): Promise<void> {
+    const texts: string[] = documents.map((doc) => doc.pageContent);
+    const vectors: number[][] = await this.embeddings.embedDocuments(texts);
+    return this.addVectors(vectors, documents);
+  }
+
+  async addVectors(vectors: number[][], documents: Document[]): Promise<void> {
+    if (vectors.length !== documents.length) {
+      throw new Error(`Vectors and metadata must have the same length`);
+    }
+    const datapoints: IndexDatapoint[] = vectors.map((vector, idx) =>
+      this.buildDatapoint(vector, documents[idx])
+    );
+    const options = {};
+    const response = await this.upsertDatapointClient.request(
+      datapoints,
+      options
+    );
+    if (Object.keys(response?.data ?? {}).length === 0) {
+      // Nothing in the response in the body means we saved it ok
+      const idDoc = documents as IdDocument[];
+      const docsToStore: Record<string, Document> = {};
+      idDoc.forEach((doc) => {
+        if (doc.id) {
+          docsToStore[doc.id] = doc;
+        }
+      });
+      await this.docstore.add(docsToStore);
+    }
+  }
+
+  // TODO: Refactor this into a utility type and use with pinecone as well?
+  cleanMetadata(documentMetadata: Record<string, any>): {
+    [key: string]: string | number | boolean | string[] | null;
+  } {
+    type metadataType = {
+      [key: string]: string | number | boolean | string[] | null;
+    };
+
+    function getStringArrays(
+      prefix: string,
+      m: Record<string, any>
+    ): Record<string, string[]> {
+      let ret: Record<string, string[]> = {};
+
+      Object.keys(m).forEach((key) => {
+        const newPrefix = prefix.length > 0 ? `${prefix}.${key}` : key;
+        const val = m[key];
+        if (!val) {
+          // Ignore it
+        } else if (Array.isArray(val)) {
+          // Make sure everything in the array is a string
+          ret[newPrefix] = val.map((v) => `${v}`);
+        } else if (typeof val === "object") {
+          const subArrays = getStringArrays(newPrefix, val);
+          ret = { ...ret, ...subArrays };
+        }
+      });
+
+      return ret;
+    }
+
+    const stringArrays: Record<string, string[]> = getStringArrays(
+      "",
+      documentMetadata
+    );
+
+    const flatMetadata: metadataType = flatten(documentMetadata);
+    Object.keys(flatMetadata).forEach((key) => {
+      Object.keys(stringArrays).forEach((arrayKey) => {
+        const matchKey = `${arrayKey}.`;
+        if (key.startsWith(matchKey)) {
+          delete flatMetadata[key];
+        }
+      });
+    });
+
+    const metadata: metadataType = {
+      ...flatMetadata,
+      ...stringArrays,
+    };
+    return metadata;
+  }
+
+  metadataToRestrictions(documentMetadata: Record<string, any>): Restriction[] {
```

Comment:
nit: We should prefer `this["FilterType"]` instead of using `Restriction[]` below on the off-chance someone wants to subclass?

Reply:
I'm trying to consider a case where that might happen, since Restriction is a type that gets passed to the API. But sure... can do that. {: